handlers/arcgis.go

```diff
@@ -0,0 +1,367 @@
+
+		metadata, err := db.ReadMetadata()
+		if err != nil {
+			return http.StatusInternalServerError, fmt.Errorf("Could not read metadata for tileset %v", id)
+		}
+
+		// TODO: pull the legend from ArcGIS specific metadata tables
+		var elements [0]interface{}
+		var layers [1]map[string]interface{}
+
+		layers[0] = map[string]interface{}{
+			"layerId":   0,
+			"layerName": toString(metadata["name"]),
+			"layerType": "",
+			"minScale":  0,
+			"maxScale":  0,
+			"legend":    elements,
+		}
+
+		out := map[string]interface{}{
+			"layers": layers,
+		}
+
+		bytes, err := json.Marshal(out)
+		if err != nil {
+			return http.StatusInternalServerError, fmt.Errorf("cannot marshal ArcGIS legend info JSON: %v", err)
+		}
+		return http.StatusOK, wrapJSONP(w, r, bytes)
+	}
+}
+
+func (s *ServiceSet) arcgisTiles(db *mbtiles.DB) handlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) (int, error) {
+		// split path components to extract tile coordinates x, y and z
+		pcs := strings.Split(r.URL.Path[1:], "/")
+		// strip off /arcgis/rest/services/ and then
+		// we should have at least <id> , "MapServer", "tiles", <z>, <y>, <x>
+		l := len(pcs)
+		if l < 6 || pcs[5] == "" {
+			return http.StatusBadRequest, fmt.Errorf("requested path is too short")
+		}
+		z, y, x := pcs[l-3], pcs[l-2], pcs[l-1]
+		tc, _, err := tileCoordFromString(z, x, y)
+		if err != nil {
+			return http.StatusBadRequest, err
+		}
+
+		var data []byte
+		err = db.ReadTile(tc.z, tc.x, tc.y, &data)
+
+		if err != nil {
+			// augment error info
+			t := "tile"
+			err = fmt.Errorf("cannot fetch %s from DB for z=%d, x=%d, y=%d: %v", t, tc.z, tc.x, tc.y, err)
+			return http.StatusInternalServerError, err
+		}
+
+		if data == nil || len(data) <= 1 {
+			// Return blank PNG for all image types
+			w.Header().Set("Content-Type", "image/png")
+			_, err = w.Write(BlankPNG())
+		} else {
+			w.Header().Set("Content-Type", db.ContentType())
+			_, err = w.Write(data)
+		}
+
+		return http.StatusOK, err
+	}
+}
+
+func geoBoundsToWMExtent(bounds []float32) arcGISExtent {
+	xmin, ymin := geoToMercator(float64(bounds[0]), float64(bounds[1]))
+	xmax, ymax := geoToMercator(float64(bounds[2]), float64(bounds[3]))
+	return arcGISExtent{
+		Xmin:             xmin,
+		Ymin:             ymin,
+		Xmax:             xmax,
+		Ymax:             ymax,
+		SpatialReference: webMercatorSR,
+	}
+}
+
+func calcScaleResolution(zoomLevel int, dpi int) (float64, float64) {
+	resolution := 156543.033928 / math.Pow(2, float64(zoomLevel))
+	scale := float64(dpi) * 39.37 * resolution // 39.37 in/m
+	return scale, resolution
+}
+
+// Cast interface to a string if not nil, otherwise empty string
+func toString(s interface{}) string {
+	if s != nil {
+		return s.(string)
+	}
+	return ""
+}
+
+// Convert a latitude and longitude to mercator coordinates, bounded to world domain.
+func geoToMercator(longitude, latitude float64) (float64, float64) {
+	// bound to world coordinates
+	if latitude > 80 {
```

Comment:
There may be a good reason for 80, I'm just curious, why is it not the usual 85? Is this an ArcGIS specialty?

Reply:
No special reason, was probably 80 in the example I used to create this...  Good catch!